In the logistic regression model, from sklearn, has a parameter C that controls regularization,which affects the complexity of the model.

The best value is dependent on the data used to train the model.

One method is to try out different values and then pick the value that gives the best score. This technique is known as a grid search

Using Default Parameters

In [1]:
from sklearn import datasets
iris = datasets.load_iris()

In [2]:
X = iris['data']
y = iris['target']

In [3]:
from sklearn.linear_model import LogisticRegression

In [4]:
logit = LogisticRegression(max_iter = 10000)

After we create the model, we must fit the model to the data.

In [5]:
print(logit.fit(X,y))

LogisticRegression(max_iter=10000)


To evaluate the model we run the score method.

In [6]:
print(logit.score(X,y))

0.9733333333333334


With the default setting of C = 1, we achieved a score of 0.973.

Let's see if we can do any better by implementing a grid search with difference values of 0.973.

In [7]:
C = [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]

In [8]:
scores = []

Loop over the range of values and update the parameter at each time

In [9]:
for choice in C:
  logit.set_params(C=choice)
  logit.fit(X, y)
  scores.append(logit.score(X, y))

In [10]:
print(scores)

[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98, 0.98, 0.9866666666666667, 0.9866666666666667]


We can see that the lower values of C performed worse than the base parameter of 1. However, as we increased the value of C to 1.75 the model experienced increased accuracy.

It seems that increasing C beyond this amount does not help increase model accuracy.